File: labs/inheritance/sorting/solution/sorting/sorting/person.cpp
#include "person.h"

std::string Person::getFirstName() const
{
	return this->firstName;
}

std::string Person::getLastName() const
{
	return this->lastName;
}

double Person::getWeightInKg() const
{
	return this->weightInKg;
}

unsigned Person::getHeightInCm() const
{
	return this->heightInCm;
}

std::ostream& operator <<(std::ostream& out, const Person& person)
{
	return out << person.getFirstName() << " " << person.getLastName() << " (weight = " << person.getWeightInKg() << "kg, height = " << person.getHeightInCm() << "cm)";
}

comparison_result Person::Comparison::ByFirstName::compare(const Person& p1, const Person& p2) const
{
	// Create local constant references
	// Allows for copy paste (better solutions are possible using templates)
	const auto& x = p1.getFirstName();
	const auto& y = p2.getFirstName();

	if (x < y) return comparison_result::LESS;
	else if (x > y) return comparison_result::GREATER;
	else return comparison_result::EQUAL;
}

comparison_result Person::Comparison::ByLastName::compare(const Person& p1, const Person& p2) const
{
	const auto& x = p1.getLastName();
	const auto& y = p2.getLastName();

	if (x < y) return comparison_result::LESS;
	else if (x > y) return comparison_result::GREATER;
	else return comparison_result::EQUAL;
}

comparison_result Person::Comparison::ByWeight::compare(const Person& p1, const Person& p2) const
{
	const auto& x = p1.getWeightInKg();
	const auto& y = p2.getWeightInKg();

	if (x < y) return comparison_result::LESS;
	else if (x > y) return comparison_result::GREATER;
	else return comparison_result::EQUAL;
}

comparison_result Person::Comparison::ByHeight::compare(const Person& p1, const Person& p2) const
{
	const auto& x = p1.getHeightInCm();
	const auto& y = p2.getHeightInCm();

	if (x < y) return comparison_result::LESS;
	else if (x > y) return comparison_result::GREATER;
	else return comparison_result::EQUAL;
}

comparison_result ComparisonInverter::compare(const Person& p1, const Person& p2) const
{
	return this->comparer->compare(p2, p1);
}

comparison_result LexicoComparer::compare(const Person& p1, const Person& p2) const
{
	auto result = c1->compare(p1, p2);

	if (result != comparison_result::EQUAL)
	{
		return result;
	}
	else
	{
		return c2->compare(p1, p2);
	}
}

File: labs/inheritance/sorting/java/src/Person.java
public class Person
{
    private String firstName;

    private String lastName;

    private double weightInKg;

    private int heightInCm;

    public Person( String firstName, String lastName, double weightInKg, int heightInCm )
    {
        this.firstName = firstName;
        this.lastName = lastName;
        this.weightInKg = weightInKg;
        this.heightInCm = heightInCm;
    }

    public String getFirstName()
    {
        return firstName;
    }

    public String getLastName()
    {
        return lastName;
    }

    public double getWeightInKg()
    {
        return weightInKg;
    }

    public int getHeightInCm()
    {
        return heightInCm;
    }

    @Override
    public String toString()
    {
        return String.format( "%s %s (weight = %fkg, height = %dcm)", firstName, lastName, weightInKg, heightInCm );
    }

    // In C++, Person.Comparison.ByFirstName becomes Person::Comparison::ByFirstName
    public static class Comparison
    {
        public static class ByFirstName implements PersonComparer
        {
            @Override
            public ComparisonResult compare( Person p1, Person p2 )
            {
                return ComparisonResult.fromInteger( p1.getFirstName().compareTo( p2.getFirstName() ) );
            }
        }

        public static class ByLastName implements PersonComparer
        {
            @Override
            public ComparisonResult compare( Person p1, Person p2 )
            {
                return ComparisonResult.fromInteger( p1.getLastName().compareTo( p2.getLastName() ) );
            }
        }

        public static class ByWeight implements PersonComparer
        {
            @Override
            public ComparisonResult compare( Person p1, Person p2 )
            {
                return ComparisonResult.fromInteger( Double.compare( p1.getWeightInKg(), p2.getWeightInKg() ) );
            }
        }

        public static class ByHeight implements PersonComparer
        {
            @Override
            public ComparisonResult compare( Person p1, Person p2 )
            {
                return ComparisonResult.fromInteger( Double.compare( p1.getHeightInCm(), p2.getHeightInCm() ) );
            }
        }
    }
}
